File: python/clase-17/final/main.py
"""Reintentos con backoff y logs estructurados para producción."""

from __future__ import annotations

import json
import os
import time
from anthropic import Anthropic, RateLimitError

MODEL = "claude-3-5-sonnet-latest"


def log_event(event: str, **fields: object) -> None:
    print(json.dumps({"event": event, **fields}, ensure_ascii=False))


def main() -> None:
    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        raise RuntimeError("Define ANTHROPIC_API_KEY.")

    client = Anthropic(api_key=api_key)
    for attempt in range(1, 4):
        try:
            response = client.messages.create(
                model=MODEL,
                max_tokens=200,
                messages=[{"role": "user", "content": "Dame un tip de observabilidad."}],
            )
            log_event("claude_response", attempt=attempt, usage=response.usage.model_dump())
            print("".join(block.text for block in response.content if block.type == "text"))
            return
        except RateLimitError:
            wait_seconds = 2 ** attempt
            log_event("rate_limited", attempt=attempt, wait_seconds=wait_seconds)
            time.sleep(wait_seconds)

    raise RuntimeError("No se pudo completar la request después de 3 intentos.")


if __name__ == "__main__":
    main()

File: typescript/clase-17/final/main.ts
import Anthropic, { RateLimitError } from "@anthropic-ai/sdk";

export {};

function logEvent(event: string, fields: Record<string, unknown>): void {
  console.log(JSON.stringify({ event, ...fields }));
}

const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) throw new Error("Define ANTHROPIC_API_KEY.");

const client = new Anthropic({ apiKey });
for (let attempt = 1; attempt <= 3; attempt += 1) {
  try {
    const response = await client.messages.create({
      model: "claude-3-5-sonnet-latest",
      max_tokens: 200,
      messages: [{ role: "user", content: "Dame un tip de observabilidad." }],
    });
    logEvent("claude_response", { attempt, usage: response.usage });
    console.log(response.content.filter((block) => block.type === "text").map((block) => block.text).join(""));
    break;
  } catch (error) {
    if (!(error instanceof RateLimitError)) throw error;
    const waitMs = 1000 * 2 ** attempt;
    logEvent("rate_limited", { attempt, waitMs });
    await new Promise((resolve) => setTimeout(resolve, waitMs));
  }
}
